backend/src/apis/snapshot_data.rs

```diff
@@ -0,0 +1,249 @@
+use num_bigint::BigInt;
+use std::collections::HashMap;
+
+use halo2_proofs::{
+    halo2curves::bn256::Fr as Fp,
+    plonk::{keygen_pk, keygen_vk},
+};
+use snark_verifier_sdk::CircuitExt;
+
+use summa_solvency::{
+    circuits::{
+        merkle_sum_tree::MstInclusionCircuit,
+        solvency::SolvencyCircuit,
+        utils::{full_prover, generate_setup_params},
+    },
+    merkle_sum_tree::utils::big_int_to_fp,
+    merkle_sum_tree::{Entry, MerkleSumTree},
+};
+
+use crate::apis::csv_parser::parse_csv_to_assets;
+
+struct SnapshotData<
+    const LEVELS: usize,
+    const L: usize,
+    const N_ASSETS: usize,
+    const N_BYTES: usize,
+    const K: u32,
+> {
+    exchange_id: String,
+    mst: MerkleSumTree<N_ASSETS>,
+    assets: Vec<Asset>,
+    proofs_of_inclusion: HashMap<u64, InclusionProof>,
+    proof_of_solvency: Option<SolvencyProof<N_ASSETS>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Asset {
+    pub name: String,
+    pub pubkeys: Vec<String>,
+    pub balances: Vec<BigInt>,
+    pub sum_balances: Fp,
+    pub signature: Vec<String>,
+}
+
+#[derive(Debug, Clone)]
+struct InclusionProof {
+    // public input
+    leaf_hash: Fp,
+    vk: Vec<u8>,
+    proof: Vec<u8>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct SolvencyProof<const N_ASSETS: usize> {
+    // public inputs
+    penultimate_node_hash: [Fp; 2],
+    assets_sum: [Fp; N_ASSETS],
+    vk: Vec<u8>,
+    proof: Vec<u8>,
+}
+
+impl<
+        const LEVELS: usize,
+        const L: usize,
+        const N_ASSETS: usize,
+        const N_BYTES: usize,
+        const K: u32,
+    > SnapshotData<LEVELS, L, N_ASSETS, N_BYTES, K>
+{
+    pub fn new(
+        exchange_id: &str,
+        entry_csv: &str,
+        asset_csv: &str,
+    ) -> Result<SnapshotData<LEVELS, L, N_ASSETS, N_BYTES, K>, Box<dyn std::error::Error>> {
+        let assets = parse_csv_to_assets(asset_csv).unwrap();
+        let mst: MerkleSumTree<N_ASSETS> = MerkleSumTree::<N_ASSETS>::new(entry_csv).unwrap();
+
+        let user_proofs = HashMap::<u64, InclusionProof>::new();
+
+        Ok(SnapshotData {
+            exchange_id: exchange_id.to_owned(),
+            mst,
+            assets,
+            proofs_of_inclusion: user_proofs,
+            proof_of_solvency: None,
+        })
+    }
+
+    fn generate_inclusion_proof(&self, user_index: usize) -> Result<InclusionProof, &'static str> {
+        // TODO: fetch pk from outside. For now, we generate them here
+        let circuit = MstInclusionCircuit::<LEVELS, L, N_ASSETS>::init_empty();
```

Comment:
I think it's ok to generate the params locally. The problem is that this way we are creating a new random `params` for each `inclusion_proof` which will then result in different `pk` and `vk`. So I think that we can safely generate the params locally, but we 
need to do it only once, maybe during `init()` and store the reuslting `pk` and `vk` somewhere in `Snapshot` before any proof generation function gets called

Reply:
I see, I created `MstParamsAndKeys` for it
Updated at [309c563](https://github.com/summa-dev/summa-solvency/pull/88/commits/309c563d3c7862aee279b07ee035095c198c3792)